### PR TITLE
Validate push descriptor/descriptor set layout create interactions

### PR DESCRIPTION
**Pull request by jzulauf-lunarg**
Add VUID checks and unit tests to address #2362, #2365.  Additionally add "required extension" check and unit test for vkCreateDescriptorSetLayout path re: VK_DESCRIPTOR_SET_LAYOUT_CREATE_PUSH_DESCRIPTOR_BIT_KHR usage, and repaired unit test not enabling required extension for ..._BIT_KHR.

VALIDATION_ERROR_05000230  VkDescriptorSetLayoutCreateInfo-flags-00280
VALIDATION_ERROR_05000232  VkDescriptorSetLayoutCreateInfo-flags-00281

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -4704,7 +4718,9 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateSampler(VkDevice device, const VkSamplerCre
 
 static bool PreCallValidateCreateDescriptorSetLayout(layer_data *dev_data, const VkDescriptorSetLayoutCreateInfo *create_info) {
     if (dev_data->instance_data->disabled.create_descriptor_set_layout) return false;
-    return cvdescriptorset::DescriptorSetLayout::ValidateCreateInfo(dev_data->report_data, create_info);
+    return cvdescriptorset::DescriptorSetLayout::ValidateCreateInfo(dev_data->report_data, create_info,
```
Comment (jzulauf-lunarg) [inline layers/core_validation.cpp]: Note the layer_data is not defined outside core_validation.cpp -- so we pass the req'd data piece-wise.

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -2173,6 +2178,15 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateDevice(VkPhysicalDevice gpu, const VkDevice
     // Store physical device properties and physical device mem limits into device layer_data structs
     instance_data->dispatch_table.GetPhysicalDeviceMemoryProperties(gpu, &device_data->phys_dev_mem_props);
     instance_data->dispatch_table.GetPhysicalDeviceProperties(gpu, &device_data->phys_dev_props);
+
+    if (device_data->extensions.vk_khr_push_descriptor) {
```
Comment (jzulauf-lunarg) [inline layers/core_validation.cpp]: This is the simplest way to get *one* extension property loaded.  Will need to refactor to capture all extension limits.

File: layers/descriptor_sets.cpp [hunk 3]
```diff
@@ -107,17 +107,54 @@ cvdescriptorset::DescriptorSetLayout::DescriptorSetLayout(const VkDescriptorSetL
...
-                                                              const VkDescriptorSetLayoutCreateInfo *create_info) {
+bool cvdescriptorset::DescriptorSetLayout::ValidateCreateInfo(const debug_report_data *report_data,
+                                                              const VkDescriptorSetLayoutCreateInfo *create_info,
+                                                              const bool push_descriptor_ext, const uint32_t max_push_descriptors) {
     bool skip = false;
     std::unordered_set<uint32_t> bindings;
+    uint64_t total_descriptors = 0;
+
+    const bool push_descriptor_set = create_info->flags & VK_DESCRIPTOR_SET_LAYOUT_CREATE_PUSH_DESCRIPTOR_BIT_KHR;
+    if (push_descriptor_set && !push_descriptor_ext) {
```
Comment (jzulauf-lunarg) [inline layers/descriptor_sets.cpp]: There is no VUID for using extension defined types, enums etc, except for the body text 

> (31.2. Extensions) ... Enabling an extension does not change behavior of functionality exposed by the core Vulkan API or any other extension, other than making valid the use of the commands, enums and structures defined by that extension.

Which implies that using the define types, etc. is *invalid* if the extension is *not* enabled..

File: scripts/helper_file_generator.py [hunk 4]
```diff
@@ -1142,22 +1143,35 @@ def GenerateTypeMapHelperHeader(self):
             '    }}',
             '    return found;',
             '}}',
+            '',
```
Comment (jzulauf-lunarg) [inline scripts/helper_file_generator.py]: helpers that allow for terse "correct by design" initialization of sType /pNext fields

File: scripts/helper_file_generator.py [hunk 5]
```diff
@@ -1175,6 +1189,13 @@ def GenerateTypeMapHelperHeader(self):
             if item.ifdef_protect != None:
                 code.append('#endif // %s' % item.ifdef_protect)
 
+        # Generate utilites for all types
```
Comment (jzulauf-lunarg) [inline scripts/helper_file_generator.py]: as the utilities reference the traits, move the utilities after the traits

File: tests/layer_validation_tests.cpp [hunk 6]
```diff
@@ -21472,6 +21472,91 @@ TEST_F(VkLayerTest, DuplicateDescriptorBinding) {
...
+    ds_layout_ci.bindingCount = 1;
+    ds_layout_ci.pBindings = &binding;
+
+    std::string error = "Attemped to use VK_DESCRIPTOR_SET_LAYOUT_CREATE_PUSH_DESCRIPTOR_BIT_KHR in ";
+    error = error + "VkDescriptorSetLayoutCreateInfo::flags but its required extension ";
+    error = error + VK_KHR_PUSH_DESCRIPTOR_EXTENSION_NAME;
+    error = error + " has not been enabled.";
+
+    m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT, error.c_str());
+    m_errorMonitor->SetDesiredFailureMsg(VK_DEBUG_REPORT_ERROR_BIT_EXT, VALIDATION_ERROR_05000232);
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: This silences an unexpected, as the second error message is implied by the first.

File: scripts/helper_file_generator.py (hunk 5 above)
Comment (mark-lunarg) [reply scripts/helper_file_generator.py]: three I's in utilities

**Comment by jzulauf-lunarg** [comment]
@tobine -- Needed to add  VK_KHR_push_descriptor to the mock_icd ignored list, because a DevicePhysicalProperties2 -> pNext query was failing to return a valid value, and causing validation to create unexpected errors in Positive tests and in the required "positive" functionality to set up negative tests.

If you have a better short term solution, I'm game.  Long term it would nice to be able to enable the extension with proper Get*Property* support.

**Comment by tobine** [comment]
@jzulauf-lunarg Yes, I'm ok with disabling VK_KHR_push_descriptor in mock for short-term. Can you open bug on me to update mock to improve GPDP2 support for extension in the mock?

**Pull request merged**